### ISSUE TITLE
Tcnative init failed with org.apache.tomcat.jni.Library.version(I)I in OSGi

### ISSUE THREAD

**Issue by xsir** [netty/netty#5539]
I am working a OSGi project with netty-tcnative + TLS and encounter a java.lang.UnsatisfiedLinkError in OpenSsl initialization. 

Here is the log:
java.lang.UnsatisfiedLinkError: org.apache.tomcat.jni.Library.version(I)I
	at org.apache.tomcat.jni.Library.version(Native Method)
	at org.apache.tomcat.jni.Library.initialize(Library.java:176)
	at io.netty.handler.ssl.OpenSsl.initializeTcNative(OpenSsl.java:243)
	at io.netty.handler.ssl.OpenSsl.<clinit>(OpenSsl.java:76)

We running a Apache Karaf 4.0.5, and here is the subset of dependent bundles:
30 | Active |  80 | 4.1.1.Final    | Netty/Codec
31 | Active |  80 | 4.1.1.Final    | Netty/Handler
32 | Active |  80 | 4.1.1.Final    | Netty/Resolver
34 | Active |  80 | 4.1.1.Final    | Netty/Common
35 | Active |  80 | 4.1.1.Final    | Netty/Transport
37 | Active |  80 | 4.1.1.Final    | Netty/Buffer
39 | Active |  80 | 1.1.33.Fork19  | Netty/TomcatNative [BoringSSL - Static]

BTW, the netty-tcnative(bundle 39) has the Bundle-NativeCode manifest entry added, according to the commit [ea27386](https://github.com/netty/netty-tcnative/commit/ea27386540f857032873292183f1c35e0c0cdbc6).

And the system information:
$ uname -a
Linux work.xsir 4.4.0-28-generic #47-Ubuntu SMP Fri Jun 24 10:09:13 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
$ java -version
java version "1.8.0_91"
Java(TM) SE Runtime Environment (build 1.8.0_91-b14)
Java HotSpot(TM) 64-Bit Server VM (build 25.91-b14, mixed mode)

I'm trying to debug the initialization of OpenSsl and netty-tcnative code. If I figured it out, I'll submit a PR ASAP.

**Comment by xsir** [comment]
BTW, I have tried the 4.1.3.Final-SNAPSHOT version, it came out the same error too.

**Comment by normanmaurer** [comment]
@xsir let us know... I have not much knowledge in terms of osgi.

**Comment by xsir** [comment]
I have figured it out finally. The root cause is: the `netty-tcnative` library is loaded by a ClassLoader that is fully independent of its bundle ClassLoader.

Specifically, the `netty-tcnative` is a bundle, which is loaded by OSGi framework with a separated ClassLoader. And the `netty-handler` is a bundle too, which contains the `OpenSsl` utility class. When the `OpenSsl` class is accessed, the static initialization will try to load the `netty-tcnative` library, which is embedded in `netty-tcnative` bundle. The OpenSsl loads the `tcnative` library through the `NativeLibraryLoader.loadFirstAvailable()` helper method, which will try to find the library in **PATH**, then in the JAR if failed. As we know, the library file is embedded in the bundle JAR, the `NativeLibraryLoader` will extract it to a tmp file, then call `System.load()` with the tmp file absolute path. The point is: in the SUN implementation, both `[System.load()](http://docs.oracle.com/javase/7/docs/api/java/lang/System.html#load(java.lang.String))` and `System.loadLibrary() ` are **CallerSensitive**, that means the native library is loaded by the caller class's ClassLoader! That's it, the `netty-tcnative` library is loaded by `NativeLibraryLoader.class`'s ClassLoader -- the netty-comon bundle's ClassLoader. But when the native method is called, it using the method definition class's ClassLoader, it can NOT find the native library in itself, then the `java.lang.UnsatisfiedLinkError: org.apache.tomcat.jni.Library.version(I)I` will be threw.

Yes, the root cause is the abused ClassLoader, it is discussed by @jjpp in [netty-tcnative issue 136](https://github.com/netty/netty-tcnative/issues/136). 

Then, the solution is clear: try to use the same ClassLoader to load the `netty-tcnative` library as the ClassLoader to initialize the native method. 

I hope I have made myself clear:) I'll submit a PR soon, thank you.
